hash-parser/src/error.rs

```diff
@@ -0,0 +1,69 @@
+//! Compiler error reporting
+//
+// All rights reserved 2021 (c) The Hash Language authors
+
+use crate::{grammar::Rule, location::Location};
+use std::fmt;
+
+/// Error message prefix
+const ERR: &str = "\x1b[31m\x1b[1merror\x1b[0m";
+
+/// Hash ParseError enum represnting the variants of possible errors.
+#[derive(Debug, Clone)]
+pub enum ParseError {
+    IoError {
+        filename: String,
+    },
+    Parsing {
+        positives: Vec<Rule>,
+        negatives: Vec<Rule>,
+        location: Location,
+    },
+    AstGeneration {
+        rule: Rule,
+        location: Location,
+    },
+}
+
+/// Convert a [pest::error::Error] into a [ParseError]
+impl<'a> From<pest::error::Error<Rule>> for ParseError {
+    fn from(pest: pest::error::Error<Rule>) -> Self {
+        match pest.variant {
+            pest::error::ErrorVariant::ParsingError {
+                ref positives,
+                ref negatives,
```

Comment:
Don't use `ref`, since you want to move them into `ParseError::Parsing`

Reply:
Could you elaborate?

Reply:
`ref` essentially gets a reference to those fields, so the type of `positives`/`negatives` will be `&Vec<Rule>`. However, since you don't want to keep the old `pest::error::Error` object, you can *move* the fields `positives` and `negatives` out of it instead of getting a reference to them. Without `ref`, the type of `positives`/`negatives` is `Vec<Rule>`, which you can immediately pass into `ParseError::Parsing` without having to call `.clone()` (thereby avoiding copying all the elements).